### PR TITLE
Improve ZFS send estimation by using actual object count

### PR DESCRIPTION
**Pull request by allanjude**
Signed-off-by: Allan Jude <allanjude@freebsd.org>

### Motivation and Context
It is desirable to be able to estimate the size of a replication stream
before sending it. This is used by various tools such as zxfer and
zrepl to provide a progress indicator (15% complete, X minutes
remaining).

Previously ZFS adjusted the estimated size of a send stream by
dividing the size of the send stream by the recordsize or volblocksize
of the dataset to calculate an approximate number of objects in the
stream, then subtracted the size of a blkptr_t for each object, and
added the size of a dmu_replay_record_t for each object.

### Description

With this change, we divide the size of the dataset by the actual
number of objects from the fill count field of the dataset root
blockpointer, to get the average object size. Then the estimated
size of the send stream is divided by the average block size, to
get an more accurate estimate of the number of objects the send
stream contains. The calculated number of dmu_replay_record_t
records is three times the number of objects, to match the contents
of the send stream.

### How Has This Been Tested?

Created a pool, single vdev, ashift=9
Created a new dataset, compress=on
Populated the dataset with 100,000 0 byte files (seq/touch)
The replication size estimate is wrong, 
```
# zfs list dirty/zerofiles
NAME              USED  AVAIL     REFER  MOUNTPOINT
dirty/zerofiles  35.5M  12.2G     35.4M  /dirty/zerofiles

# zfs send -vP dirty/zerofiles@100000 | zstreamdump
full    dirty/zerofiles@100000  69869264 (66.6M)
size    69869264

BEGIN record
        hdrtype = 1
        features = 4
        magic = 2f5bacbac
        creation_time = 5c6b51c6
        type = 2
        flags = 0x4
        toguid = 156b5fd451052d15
        fromguid = 0
        toname = dirty/zerofiles@100000
END checksum = 5b128c2b84ecc6/42d9c8b5392f8d1a/4a3a82d9d2c3450d/19395d1ca1b44c72
SUMMARY:
        Total DRR_BEGIN records = 1
        Total DRR_END records = 1
        Total DRR_OBJECT records = 300006
        Total DRR_FREEOBJECTS records = 12
        Total DRR_WRITE records = 200519
        Total DRR_WRITE_BYREF records = 0
        Total DRR_WRITE_EMBEDDED records = 0
        Total DRR_FREE records = 400012
        Total DRR_SPILL records = 0
        Total records = 900551
        Total write size = 110840832 (0x69b4c00)
        Total stream length = 443012912 (0x1a67d730)
```

The send estimation was too low by 534%

In another example, using a FreeBSD 13 source tree:
```
# zfs list dirty/src
NAME        USED  AVAIL     REFER  MOUNTPOINT
dirty/src  1.26G  12.2G     1.25G  /dirty/src

# zfs send -vP dirty/src@src | zstreamdump
full    dirty/src@src   1319144416 (1.23G)
size    1319144416
BEGIN record
        hdrtype = 1
        features = 4
        magic = 2f5bacbac
        creation_time = 5c663d3e
        type = 2
        flags = 0x4
        toguid = b849de70be45a523
        fromguid = 0
        toname = dirty/src@src
END checksum = 568a3fd73cecb08/54621f0599a442a4/548696cf599f76f9/c4134c6715ce6c48
SUMMARY:
        Total DRR_BEGIN records = 1
        Total DRR_END records = 1
        Total DRR_OBJECT records = 254667
        Total DRR_FREEOBJECTS records = 13
        Total DRR_WRITE records = 257513
        Total DRR_WRITE_BYREF records = 0
        Total DRR_WRITE_EMBEDDED records = 0
        Total DRR_FREE records = 256195
        Total DRR_SPILL records = 0
        Total records = 768390
        Total write size = 1346616832 (0x5043be00)
        Total stream length = 1629816824 (0x612507f8)
```

Estimate was too small by 23.5%

After my patch:
100000x 0 byte files: Estimate: 298M, Actual: 443M, too small by 41.6%
FreeBSD src tree: Estimate: 1.42G, Actual: 1.51G, too small by 7%

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/dmu_send.c [hunk 1]
```diff
@@ -1378,8 +1379,21 @@ dmu_adjust_send_estimate_for_indirects(dsl_dataset_t *ds, uint64_t uncompressed,
...
 	} else {
-		err = dsl_prop_get_int_ds(ds,
-		    zfs_prop_to_name(ZFS_PROP_RECORDSIZE), &recordsize);
+		/*
+		 * Rather than using the value of the recordsize property of the
+		 * dataset, get the number of live objects in the dataset, and
+		 * its uncompressed size, and use those to calculate the
+		 * average recordsize. Then use that value and the size of
+		 * the snapshot we are sending to calculate a more accurate
+		 * record_count.
```
Comment (tcaputi) [inline module/zfs/dmu_send.c]: a more accurate size?

**Comment by ahrens** [comment]
> Previously ZFS adjusted the estimated size of a send stream by
> dividing the size of the send stream by the recordsize or volblocksize
> of the dataset to calculate an approximate number of objects

That math is to calculate the number of *blocks*, not objects. The fill count of the root BP (i.e. the number of objects aka allocated dnodes) is not related to the recordsize.

It might make sense to estimate the number of changed objects, and use this to add in the overhead for OBJECT records, in addition to the code that estimates the number of changed blocks and adds the overhead for WRITE records.  (Although for incrementals, estimating the number of modified objects seems very workload-dependent.)

**Comment by allanjude** [comment]
Matt and I have discussed what changes need to be made, and I'll update this pull request in the next few days

**Comment by Ornias1993** [comment]
@allanjude Did you forget to push the update, or are you leaving this PR be for now?

**Comment by ahrens** [comment]
I noticed that this PR hasn’t been updated in some time.  We would like to see this feature added, but it seems that it isn’t quite complete, so we’re going to close it for now.  When you have time to revisit this work, feel free to reopen this PR or open a new one.  Thanks for your contribution to OpenZFS and please continue to open PR’s to address issues that you encounter!

**Pull request closed without merging**